### ISSUE TITLE
Support of dynamic dimensions.

### ISSUE THREAD

**Issue by strasdat** [coreylowman/dfdx#84]
https://www.reddit.com/r/rust/comments/vy5uwf/announcing_dfdx_an_deep_learning_library_built/ig0iwlv?utm_medium=android_app&utm_source=share&context=3

To be specific, this feature request is NOT about making the rank of a tensor dynamic. It is rather to have the option of setting one or more dimensions in a shape to "dynamic".

**Comment by strasdat** [comment]
For example, how this is done in the nalgebra crate: https://docs.rs/nalgebra/latest/nalgebra/base/dimension/trait.Dim.html

**Comment by jafioti** [comment]
@strasdat Is this compatible with const generic dims? It seems to be a trait, so I'm not sure how you could have a generic const generic.

**Comment by strasdat** [comment]
Yeah, I looked into it a little more. nalgebra does a hack using "type style integers" and not const generics. Supporting this through const generics might be tricky or even impossible until some sort of const generics specialization is stabilized.

Let's look at an example - assuming that "0" has a special meaning and indicates a dynamic dimensions:

For e.g. Tensor2D<5, 0> we would like to fail at compile-time if the first dimension (5) does not match for a corresponding operation. But we would like to check at run-time if the second dimension (which is set  at run-time) matches.

Hence, we would need a specialization (for 0). 

Seems like this feature might have to wait until rust catches up and we might be stuck with either nalgebra approach, all dynamic (as in ndarray) or all at compile time (as in dfdx) - unless somebody comes up with a creative workaround.

**Comment by strasdat** [comment]
One possible workaround could be to define a lot of structs:

Rank 1 tensors:

TensorR1<M> - static
TensorR1_d - dynamic

Rank 2 tensors:
TensorR2<M,N> - all static
TensorR2_ds <N> - first axis dynamic
TensorR2 _sf<M> - second axis dynamic
TensorR2_dd - all axes dynamic

Rank 3 tensors:
TensorR3<M,N,O> - all static
TensorR3_dss <N,O> - first axis dynamic
TensorR3 _sds<M,O> - second axis dynamic
TensorR3 _ssd<M,N> - third axis dynamic
[...]
TensorR3_ddd - all axes dynamic

This gets a bit combinatorial at higher rank tensors, but one could also just implement a subset, e.g. for 4 rank tensor just:

TensorR4<M,N,O> - all static

**Comment by coreylowman** [comment]
Yeah for const generic version we'd probably end up with something like:

```rust
trait Dim {
    fn value(&self) -> usize;
}

struct CompileTimeDim<const N: usize>;
impl<const N: usize> Dim for CompileTimeDim<N> { ... }

struct RunTimeDim(usize);
impl Dim for RunTimeDim { ... }

struct Tensor2D<const D1: Dim, const D2: Dim> {}
```

But as you note something this would require trait bounds on const variables, and also being able to use custom structs in const variables, not just integers. So probably need a bit more support.

Will have to think about what that means for the underlying data storage as well. I think it'd have to be a Vec (or Slice?) for the dynamic axis

**Comment by strasdat** [comment]
> Will have to think about what that means for the underlying data storage as well. I think it'd have to be a Vec (or Slice?) for the dynamic axis

Yeah I believe we would want static allocations with arrays as if all axes are static and some kind of heap allocation like Vec if at least one axis and hence the overall size is dynamic.

**Comment by jafioti** [comment]
I posted this in another issue, but I'll put it here so all ideas are in one place. In my experimental dl crate `condor` I approached this using enums:
```rust
enum DimType {
    Static(usize),
    Dynamic,
}
```
The main downside is that you can't specify equality correctly yet, for instance you should be able to specify Static == Dynamic, Dynamic == Static, but Static(a) != Static(b) if a != b. Currently you can only derive PartialEq for any types used as const generics, so you can't implement custom equality logic.

**Comment by strasdat** [comment]
So yeah, googled a little and found this older issue of yours (https://github.com/rust-lang/rust/issues/94661).

It seems like PartialEq won't work here, because the relationship is not transitive: Static(a)==Dynamic , Static(b)==Dynamic does not imply Static(a)==Static(b).

But I like the idea of using such enum and maybe one could define a custom Trait here. I'll experiment a little... (I have not done much with const generics in rust...)
